cuda/trsv_kernels.cuh

```diff
@@ -0,0 +1,1035 @@
+            const auto rhs_val = x[row * x_info.stride];
+            x[row * x_info.stride] = (rhs_val - local_res) / diag;
+        }
+    }
+}
+
+
+namespace kernel {
+
+
+// Must be called with 2-D blocks and a 1-element grid
+template <std::int64_t diag_block_size, dmtx_t dmtx, typename ValueType>
+__global__ __launch_bounds__(diag_block_size *diag_block_size) void lower_trsv(
+    const std::int64_t diag_block_start, const matrix_info m_info,
+    const ValueType *__restrict__ mtx, const matrix_info x_info,
+    ValueType *__restrict__ x)
+{
+    // assert: blockDim.x == blockDim.y == diag_block_size
+    // assert: blockIdx.x == blockIdx.y == 1
+    const std::int64_t row_idx = diag_block_start + threadIdx.x;
+    const std::int64_t col_idx = diag_block_start + threadIdx.y;
+    if (diag_block_start >= m_info.size[0]) {
+        return;
+    }
+    const auto group = cg::this_thread_block();
+    constexpr int shared_stride = diag_block_size + 1;
+    __shared__ ValueType shared[diag_block_size * shared_stride];
+
+    // Read into shared memory all necessary matrix data and store it transposed
+    // (with stride = BS+1)
+    if (threadIdx.x <= threadIdx.y && row_idx < m_info.size[0] &&
+        col_idx < m_info.size[1]) {
+        shared[threadIdx.x * shared_stride + threadIdx.y] =
+            (dmtx == dmtx_t::unit && threadIdx.x == threadIdx.y)
+                ? ValueType{1}
+                : mtx[row_idx + col_idx * m_info.stride];
+    }
+    group.sync();
+    auto warp = cg::tiled_partition<WARP_SIZE>(group);
+    if (group.thread_rank() / WARP_SIZE == 0) {
+        const auto warp_id = warp.thread_rank();
+        const std::int64_t x_idx = (diag_block_start + warp_id) * x_info.stride;
+        auto local_solution = x[x_idx];
+
+        for (int i = 0; i < WARP_SIZE; ++i) {
+            const auto mtx_val = shared[i * shared_stride + warp_id];
+            const auto current_x = warp.shfl(local_solution / mtx_val, i);
+            if (warp_id == i) {
+                local_solution = current_x;
+            }
+            if (warp_id > i) {
+                local_solution -= mtx_val * current_x;
+            }
+        }
+        // Write solution back
+        if (diag_block_start + warp_id < x_info.size[0]) {
+            x[x_idx] = local_solution;
+        }
+    }
+}
+
+// Must be called with 1-D blocks and 1-D grid
+template <std::int64_t block_size, std::int64_t solved_size, typename ValueType>
+__global__ __launch_bounds__(block_size) void lower_trsv_apply(
+    const std::int64_t apply_start, const matrix_info m_info,
+    const ValueType *__restrict__ mtx, const matrix_info x_info,
+    ValueType *__restrict__ x)
+{
+    // assert: blockDim.x == blockDim.y == diag_block_size
+    // assert: blockDim.x % solved_size == 0
+    const std::int64_t global_id = blockIdx.x * blockDim.x + threadIdx.x;
+    const std::int64_t mtx_row =
+        apply_start + solved_size + global_id / solved_size;
```

Comment:
if the solved_size is not the multiple of warp size, the warp in the following will not be correct